src/snowflake/cli/api/commands/decorators.py

```diff
@@ -136,10 +146,15 @@ def _options_decorator_factory(
     additional_options: List[inspect.Parameter],
     execute_before_command_using_new_options: Optional[Callable] = None,
 ):
+    """
+    execute_before_command_using_new_options executes before command telemetry has been emitted,
+    but after command line options have been populated.
+    """
+
     @wraps(func)
     def wrapper(**options):
         if execute_before_command_using_new_options:
-            execute_before_command_using_new_options()
+            execute_before_command_using_new_options(**options)
```

Comment:
Should it be `_execute_before_command_using_global_options`? @sfc-gh-pjob wdyt about using this for registering the callbacks?

Reply:
it's not restricted to global options. it gets all the parameters of the current command.

Reply:
@sfc-gh-turbaszek I'm not sure what exactly is the subject of your question.

Maybe I miss sth in the changes but I think changes in this function `_options_decorator_factory` are only cleanup changes and aren't responsible for the effect of this PR. @sfc-gh-melnacouzi please correct me if I'm wrong. And I think these changes are good changes but maybe not necessary should be part of this PR (but IMO they can be).

@sfc-gh-turbaszek I think all the magic goes via `typer_pre_execute.py` and I'm not sure about keeping functions to run before commands in our context but IMO it looks OK and simple to use.

Reply:
My concern with this approach is the implicit order of callbacks. Callbacks will be registered in the order of parsing/evaluating commands, this is a very deep implementation detail imho.